components/api/src/test/java/com/hotels/styx/api/ByteStreamTest.java

```diff
@@ -0,0 +1,185 @@
+    private Buffer buf3;
+
+    @BeforeMethod
+    public void setUp() {
+        buf1 = new Buffer("a", UTF_8);
+        buf2 = new Buffer("b", UTF_8);
+        buf3 = new Buffer("c", UTF_8);
+    }
+
+    @Test
+    public void publishesContent() {
+        ByteStream stream = new ByteStream(Flux.just(buf1, buf2, buf3));
+
+        StepVerifier.create(stream)
+                .expectNext(buf1)
+                .expectNext(buf2)
+                .expectNext(buf3)
+                .verifyComplete();
+    }
+
+    @Test
+    public void publisherBackpressure() {
+        ByteStream stream = new ByteStream(Flux.just(buf1, buf2, buf3));
+
+        StepVerifier.create(stream, 0)
+                .expectSubscription()
+                .thenRequest(1)
+                .expectNext(buf1)
+                .thenRequest(2)
+                .expectNext(buf2, buf3)
+                .verifyComplete();
+    }
+
+    @Test
+    public void mapsContent() {
+        ByteStream stream = new ByteStream(Flux.just(buf1, buf2, buf3));
+
+        ByteStream mapped = stream.map(this::toUpperCase);
+
+        StepVerifier.create(Flux.from(mapped).map(this::decodeUtf8String))
+                .expectSubscription()
+                .expectNext("A", "B", "C")
+                .verifyComplete();
+    }
+
+    @Test
+    public void releasesRefcountForMappedBuffers() {
+        ByteStream stream = new ByteStream(Flux.just(buf1, buf2, buf3));
+
+        ByteStream mapped = stream.map(this::toUpperCase);
+
+        StepVerifier.create(Flux.from(mapped).map(this::decodeUtf8String))
+                .expectSubscription()
+                .expectNext("A", "B", "C")
+                .verifyComplete();
+
+        assertThat(buf1.delegate().refCnt(), is(0));
+        assertThat(buf2.delegate().refCnt(), is(0));
+        assertThat(buf3.delegate().refCnt(), is(0));
+    }
+
+    @Test
+    public void mapRetainsRefcountsForInlineBufferChanges() {
+        ByteStream stream = new ByteStream(Flux.just(buf1, buf2, buf3));
+
+        ByteStream mapped = stream.map(buf -> buf);
+
+        StepVerifier.create(Flux.from(mapped).map(this::decodeUtf8String))
+                .expectSubscription()
+                .expectNextCount(3)
+                .verifyComplete();
+
+        assertThat(buf1.delegate().refCnt(), is(1));
+        assertThat(buf2.delegate().refCnt(), is(1));
+        assertThat(buf3.delegate().refCnt(), is(1));
+    }
+
+    @Test
+    public void discardsContent() {
+        ByteStream stream = new ByteStream(Flux.just(buf1, buf2, buf3));
+
+        ByteStream discarded = stream.drop();
+
+        StepVerifier.create(discarded)
+                .expectSubscription()
+                .verifyComplete();
+
+        assertThat(buf1.delegate().refCnt(), is(0));
+        assertThat(buf2.delegate().refCnt(), is(0));
+        assertThat(buf3.delegate().refCnt(), is(0));
+    }
+
+    @Test
+    public void aggregatesContent() throws ExecutionException, InterruptedException {
```

Comment:
We should also have a test for when the aggregation goes over the maximum bytes.